Support Running Unit Tests With `wp-env`

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Instead of requiring contributors to set up an environment for unit tests, we should provide support for running them through `wp-env`. This provides a simple command that can be used to run the unit tests. As a note, since this is a separate command, it _currently_ doesn't do anything unless you use it on purpose. I've documented it here as being experimental, so, we can continue to work on it both in our repository and upstream in `@wordpress/env`.

Contributes to https://github.com/woocommerce/woocommerce/issues/32436.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. [Make sure you've set up `wp-env`](https://github.com/woocommerce/woocommerce/blob/trunk/DEVELOPMENT.md#plugin-development-environments)
2. Try `pnpm --filter=woocommerce test:unit:env` locally
3. Ensure all tests pass
4. Try passing PHPUnit arguments, such as `pnpm --filter=woocommerce test:unit:env --filter=test_handle`

<!-- End testing instructions -->

ObliviousHarmony committed:
Updated `@wordpress/env`

ObliviousHarmony committed:
Added `wp-env` PHPUnit Support

This commit adds a new `test:unit:env` script to the WooCommerce
plugin. This runs PHPUnit from within `wp-env` and keeps you from
having to set up a testing environment for the unit tests.

ObliviousHarmony committed:
Documented `wp-env` Unit Test Usage

ObliviousHarmony committed:
Changelogs

psealock commented on plugins/woocommerce/.wp-env.json:
```diff
@@ -16,7 +16,12 @@
             "port": 8086,
+            "plugins": [
+                ".",
+                "https://downloads.wordpress.org/plugin/akismet.zip"
```
TIL this downloads the latest when a version is not supplied

psealock commented on plugins/woocommerce/tests/README.md:
```diff
@@ -5,19 +5,24 @@ This document discusses unit tests. See [the e2e README](https://github.com/wooc
+## Recommended Setup
 
-## Initial Setup
+We provide support for running unit tests within a `wp-env` container. All you have to do is complete the [Getting Started Guide](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce#getting-started) and have a functional `wp-env` environment.
```
>All you have to do is

Is this a bit reductive? 

>and have a functional `wp-env` environment

I don't thing there are steps for this, like `npm -g i @wordpress/env`, so maybe reword a bit.

```suggestion
Complete the [Getting Started Guide](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce#getting-started) and you will have a functional `wp-env` environment to run unit tests.
```

ObliviousHarmony committed:
Update plugins/woocommerce/tests/README.md

Co-authored-by: Paul Sealock <psealock@gmail.com>

psealock commented on plugins/woocommerce/tests/README.md:
```diff
@@ -53,19 +58,14 @@ Example:
+The tests will execute and you will be presented with a summary.
 
-    $ vendor/bin/phpunit --coverage-text
+You can run specific test files by providing the path to the test class, such as `pnpm --filter=woocommerce test:unit:env tests/legacy/unit-tests/importer/product.php`. A text code coverage summary can be displayed using the `--coverage-text` option.
```
Is this the same as the testing instructions?

>pnpm --filter=woocommerce test:unit:env --filter=test_handle

There is no `--filter` bit.

Running `pnpm --filter=woocommerce test:unit:env tests/legacy/unit-tests/importer/product.php` produces an error for me:

```
Cannot open file "tests/legacy/unit-tests/importer/product.php".
```

ObliviousHarmony replied:
For the `--filter` it should run 10 tests instead of the thousands. You're passing `--filter=test_handle` to PHPUnit which then, well, filters.

As for the file path, you're right, that doesn't seem to work. I think it's because the working directory might not be great, let me look into that.

ObliviousHarmony replied:
It looks like the path to a specific file isn't going to work right now. We're going to need an upstream PR to set the working directory when using `wp-env run` 😓 If we brand this as experimental I think we can still keep it. It isn't used anywhere except manually.

ObliviousHarmony committed:
Isolated PHPUnit Database Tables

By default it was using the standard table prefix and
destroyed the test instance as a result.

ObliviousHarmony committed:
Clarify Experimental Nature

Since there's still some upstream work it's important to note
that this is an experimental feature!

MrJnrman committed:
Fix E2E test (#37804)

MrJnrman commented on plugins/woocommerce/tests/e2e-pw/bin/test-env-setup.sh:
```diff
@@ -4,41 +4,45 @@ ENABLE_HPOS="${ENABLE_HPOS:-0}"
-docker-compose -f $(wp-env install-path)/docker-compose.yml run --rm -u $(id -u) -e HOME=/tmp tests-cli sh -c 'wp option update blogname "WooCommerce Core E2E Test Suite"'
+wp-env run tests-cli 'wp option update blogname "WooCommerce Core E2E Test Suite"'
+
+wp-env run tests-cli "wp plugin install https://gist.github.com/vedanshujain/564afec8f5e9235a1257994ed39b1449/archive/b031465052fc3e04b17624acbeeb2569ef4d5301.zip --activate"
```
This is a left over from testing and should be removed.

ObliviousHarmony committed:
Removed Accidental COT Check Bypass

ObliviousHarmony commented:
I've opened a few upstream pull requests to clean this PR up:

- https://github.com/WordPress/gutenberg/pull/49883: Right now the port is manually added to `WP_TESTS_DOMAIN` in the `.wp-env.json` file. This gives some weird console output and I'm worried the double port may have other implications.
- https://github.com/WordPress/gutenberg/pull/49908: Developers would expect `pnpm --filter=woocommerce run test:unit:env` to support relative paths to test files. Unfortunately, right now, paths are relative to `/var/www/html` instead of the plugin.
- https://github.com/WordPress/gutenberg/pull/49962: The E2E setup files included in this PR highlight a somewhat significant problem with the way `wp-env` deals with permissions. If we're going to suggest that people use this for their tests and development environment, we need to make sure it works well.

ObliviousHarmony commented:
A few upstream changes have invalidated some of the work in this pull request. Given that and its age, I am going to close it in favor of creating a new one once the next version of `@wordpress/env` is released. I would also like to separate this into a few pull requests.

- **Update E2E `wp-env` Version**: This should include the updates to the E2E test suite in this PR along with others to make use of the upstream changes to permissions and setup scripts.
- **Support Running Unit Tests With `wp-env`**: This will add the parts of this pull request to run the `test:unit:env` script. It should also make use of the `--env-cwd` option.
- **Use `wp-env` To Run GitHub Action Unit Tests**: The last step will be to utilize `wp-env` in the unit tests in CI.

Once we've completed these, technically, the issue will be done. I'd like to plan some follow-up work around cleaning up the scripts in `plugins/woocommerce/package.json`, but, this is a great start to supporting this new paradigm. Once we've battle-tested it we can make this the opinionated default.
